lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -1265,8 +1282,75 @@ struct InstOpConversion : public ConvertToLLVMPattern {
+          auto zeroC = initBuilder.create<LLVM::ConstantOp>(
+              op.getLoc(), i32Ty, rewriter.getI32IntegerAttr(0));
+
+          auto numElements = initBuilder.create<LLVM::ConstantOp>(
+              op.getLoc(), i32Ty,
+              rewriter.getI32IntegerAttr(underlyingTy.getArrayNumElements()));
+
+          // Get element size.
+          auto null = initBuilder.create<LLVM::NullOp>(
+              op.getLoc(), underlyingTy.getPointerTo());
+          auto gepFirst = initBuilder.create<LLVM::GEPOp>(
+              op.getLoc(), underlyingTy.getArrayElementType().getPointerTo(),
+              null, ArrayRef<Value>({zeroC, oneC}));
+          auto toInt = initBuilder.create<LLVM::PtrToIntOp>(op.getLoc(), i32Ty,
+                                                            gepFirst);
+
+          // Add information to the state.
+          initBuilder.create<LLVM::CallOp>(
+              op.getLoc(), llvm::None,
+              rewriter.getSymbolRefAttr(addSigElemFunc),
+              ArrayRef<Value>({initStatePtr, sigIndex, toInt, numElements}));
+        } else if (underlyingTy.isStructTy()) {
+          auto zeroC = initBuilder.create<LLVM::ConstantOp>(
+              op.getLoc(), i32Ty, rewriter.getI32IntegerAttr(0));
+
+          auto null = initBuilder.create<LLVM::NullOp>(
+              op.getLoc(), underlyingTy.getPointerTo());
+          for (size_t i = 0, e = underlyingTy.getStructNumElements(); i < e;
+               ++i) {
+            auto oneC = initBuilder.create<LLVM::ConstantOp>(
+                op.getLoc(), i32Ty, rewriter.getI32IntegerAttr(1));
+            auto indexC = initBuilder.create<LLVM::ConstantOp>(
+                op.getLoc(), i32Ty, rewriter.getI32IntegerAttr(i));
+
+            // Get pointer offset.
+            auto gepElem = initBuilder.create<LLVM::GEPOp>(
+                op.getLoc(),
+                underlyingTy.getStructElementType(i).getPointerTo(), null,
+                ArrayRef<Value>({zeroC, indexC}));
+            auto elemToInt = initBuilder.create<LLVM::PtrToIntOp>(
+                op.getLoc(), i32Ty, gepElem);
+
+            // Get element size.
+            auto elemNull = initBuilder.create<LLVM::NullOp>(
+                op.getLoc(),
+                underlyingTy.getStructElementType(i).getPointerTo());
+            auto gepElemSize = initBuilder.create<LLVM::GEPOp>(
+                op.getLoc(),
+                underlyingTy.getStructElementType(i).getPointerTo(), elemNull,
+                ArrayRef<Value>({oneC}));
+            auto elemSizeToInt = initBuilder.create<LLVM::PtrToIntOp>(
+                op.getLoc(), i32Ty, gepElemSize);
+
+            // Add information to the state.
+            initBuilder.create<LLVM::CallOp>(
+                op.getLoc(), llvm::None,
+                rewriter.getSymbolRefAttr(addSigStructFunc),
+                ArrayRef<Value>(
+                    {initStatePtr, sigIndex, elemToInt, elemSizeToInt}));
+          }
+        }
```

Comment:
At some point we'll want to write up some documentation explaining how exactly the runtime layout of the various LLHD data types looks like. Especially taking into account signals and any weird tricks we have to play.

But I'm very much in favor of doing this at a later point, and getting this merged ASAP to ensure all of the awesome `llhd-sim` developments are upstream.